Add type annotations

bluetech opened the pull request:
- Annotate the package.
- Annotate the tests.
- Run mypy in CI (in both Python 2 and Python 3 modes).
- Publish the types with `py.typed` files.

Fixes #339.

pquentin commented:
Thanks for tackling this!

That's a lot of ignores. :)

Can you please look into the Python 3 test failure?

pquentin commented:
I also enabled https://github.blog/changelog/2021-07-01-github-actions-new-settings-for-maintainers/ so that you won't have to wait for me to see the CI results.

bluetech commented:
> That's a lot of ignores. :)

I didn't think so. Most of the ignores are for `TypeError` tests, where they are intentional. I hope one day to come back to https://github.com/python/mypy/issues/8655 but don't get around to it.

> Can you please look into the Python 3 test failure?

So the problem here is that mypy only runs on Python 3, so installs Python 3 dependencies, specifically pytest 6, but then `mypy -2` fails with syntax errors. I'll sort it out somehow.

> I also enabled https://github.blog/changelog/2021-07-01-github-actions-new-settings-for-maintainers/ so that you won't have to wait for me to see the CI results.

Thanks!

pquentin commented:
I'm happy with mypy only working on Python 3 if that helps!

bluetech commented:
I tried some more fixes but now there are two other problems:

- `mypy[python2]` requires `typed-ast` which doesn't support PyPy.
- pytest 4.6.x doesn't support Python 3.10

Possible solutions are:

1. Have a dedicated mypy/lint job instead of running from `ci.sh` in each job. Probably a good idea anyway.
2. Drop Python 2.

Since I already made the effort I'll go for 1, but we can do 2 as well in a (separate?) major release.

pquentin commented:
I like that plan. 👍

bluetech commented:
OK, I moved mypy to a separate lint jobs and left the CI job as is, seems to work.

For the coverage I added an exclude of `if TYPE_CHECKING:` but that doesn't seem to work in CI, even though it works locally, not sure why. I'll try to add explicit `pragma: no cover` maybe that will do the trick.

bluetech commented:
Codecov timed out earlier but it's green now. I'll work on dropping Python 2 next when I get a chance. I'll leave it to you to decide whether to do a final py2 release with typing or just go straight to py3.

pquentin commented:
Thanks! I don't have a strong opinion on including Python 2 removal in the next release or not. I just might not get to it before you.

pquentin commented:
@bluetech I just released 0.9.0 with this pull request

bluetech commented:
@pquentin Thanks!

@hramezani FYI - urllib3 should be able to remove the type ignores for trustme now if it installs trustme>=0.9.0.

hramezani commented:
@bluetech Thanks for letting me know. Here is the PR https://github.com/urllib3/urllib3/pull/2371
